### DISCUSSION TITLE
Converting a record or table object to text and/or a way to access fields within a table/record that are embedded

### DISCUSSION THREAD

**Discussion by rbeste** [microsoft/Power-Fx#2624] [Q&A] [answered]
Have a Copilot Studio project utilizing the connector to Azure CLU which returns its result as a record object.  I just need to the key values out of that record so I can pass it to a Power Automate Flow which only takes text as an input but I can't seem to figure out how to get those key field(s) into a text string. JSON/PARSEJSON don't work on Record/Table objects and TEXT doesn't either.  When trying to use CONCAT I can get so far but since the structure of that record is below, how do I get to the key values since referencing them with record.prediction.entities[0].extraInformation[0].key throws an error saying it doesn't like the [0] notation.  What am I missing here?
 {
  "prediction": {
    "entities": [
      {
        "category": "Service",
        "confidenceScore": 1,
        "extraInformation": [
          {
            "extraInformationKind": "ListKey",
            "key": "value returned 1"
          },
          {
            "extraInformationKind": "ListKey",
            "key": "value returned 2"
          }
        ],
        "length": 18,
        "multipleResolutions": false,
        "offset": 0,
        "text": "CRM implementation",
        "topResolution": {
          "resolutionKind": "TextSpan",
          "value": "CRM implementation"
        }
      }
    ],
    "intents": [
      {
        "category": "Service_Query",
        "confidenceScore": 0.9256307
      },
      {
        "category": "Greeting_Query",
        "confidenceScore": 0.4805507
      },
      {
        "category": "ThankYou_Query",
        "confidenceScore": 0.4247672
      },
      {
        "category": "Goodbye_Query",
        "confidenceScore": 0.24324831
      },
      {
        "category": "StartOver_Query",
        "confidenceScore": 0.23821941
      },
      {
        "category": "None",
        "confidenceScore": 0
      }
    ],
    "projectKind": "Conversation",
    "topIntent": "Service_Query"
  },
  "query": "questionasked"
}

**Comment by anderson-joyle** [comment]
I'll need to confirm which Power-Fx version Copilot Studio is running. A recent change introduced support for the `Index` function with untyped objects.
```
>> Set(t,ParseJSON("{ ""prediction"": { ""entities""....}")
<Untyped: Use Value, Text, Boolean, or other functions to establish the type>

>> Text( Index( Index( t.prediction.entities, 1).extraInformation, 1).key)
"value returned 1"
```

Currently `Concat` does not support untyped objects. A possible way yo achieve your goal is to combine `Concat` + `ForAll`:
```
>> Set(t,ParseJSON("{ ""prediction"": { ""entities""....}")
<Untyped: Use Value, Text, Boolean, or other functions to establish the type>

>> Concat(ForAll(Index(t.prediction.entities,1).extraInformation, Text(ThisRecord.key)), Value, ",")
"value returned 1,value returned 2"
```

**Reply by rbeste** [reply, marked as answer]
I've tried Concat with ForAll I can't tell you how many iterations and it always comes back to either the field (key) is unknown or Value is seen as a Table and it doesn't like that.  After trying everything I could find, I finally got it to work using the following: But there has got to be a better way of simply accessing fields in a table/record object as strings than this...
    With(
        {
            keys: ForAll(
                Global.ConversationResult.prediction.entities As en,
                Concat(
                    ForAll(
                        en.extraInformation As ex,
                        If(
                            ex.extraInformationKind = "ListKey",
                            ex.key,
                            Blank()
                        )
                    ),
                    Value,
                    ", "
                )
            )
        },
        Concat(
            Split(keys, ", "),
            Result,
            ", "
        )
    )

**Reply by MariaGabrielaMello** [reply]
Hi, did you find any solutions? I'm currently facing the same problem and only got so far as you did...

**Reply by rbeste** [reply]
this finally worked:

ForAll(
 Global.ConversationResult.prediction.entities As en,
  If(
    !IsBlank(Coalesce(en.extraInformation, Blank())),
    ForAll(
      en.extraInformation As ex,
      If(
        !IsBlank(Coalesce(ex.extraInformationKind, Blank())) && ex.extraInformationKind = "ListKey",
        {key: ex.key}, // Keep only the key value in the table
        Blank()
      )
    ),
    Blank() // Provide an empty object if extraInformation is missing
  )
)

**Reply by MariaGabrielaMello** [reply]
Thank you for replying, I'll try to apply it to my solution!!
